arangod/Pregel/Worker/Worker.cpp

```diff
@@ -332,16 +352,37 @@ void Worker<V, E, M>::_startProcessing() {
   Scheduler* scheduler = SchedulerFeature::SCHEDULER;
 
   _feature.metrics()->pregelNumberOfThreads->fetch_add(1);
+  _initializeMessageCaches();
 
   auto self = shared_from_this();
   scheduler->queue(RequestLane::INTERNAL_LOW, [self, this] {
     if (_state != WorkerState::COMPUTING) {
       LOG_PREGEL("f0e3d", WARN) << "Execution aborted prematurely.";
       return;
     }
-    if (_processVertices() && _state == WorkerState::COMPUTING) {
-      _finishedProcessing();  // last thread turns the lights out
+
+    std::vector<futures::Future<Result>> futures;
+    // first build up all future requests
+    for (auto [idx, quiver] : enumerate(_quivers)) {
+      TRI_ASSERT(_state == WorkerState::COMPUTING);
+      futures.emplace_back(_processVertices(idx, quiver));
```

Comment:
This does not work like you think it does; With this code you run `_processVertices` sequentially and all instances return a trivial future just containing a value; nothing happens concurrently.

You need to create a future that executes `processVertices`, not return futures from `processVertices`;

So here you have to
```suggestion
      futures.emplace_back(futures::makeFuture([idx, quiver]() { return_processVertices(idx, quiver); }));
```